### PR TITLE
ci(coverage): per-crate coverage gate for objects / repo / refs

### PR DESCRIPTION
**Pull request by thethorne48**
## Summary
- New `heddle-devtools audit-coverage` subcommand parses `lcov.info` and fails per-crate when line coverage drops below a `--gate <crate>=<pct>` threshold (mirrors the existing `audit-idempotency` pattern in the same crate).
- `.github/workflows/rust-tests.yml` runs the gate after `cargo llvm-cov` emits `lcov.info` and before the Codecov upload — gate is in-CI authoritative, Codecov is the PR-reporting mirror.
- New `codecov.yml` declares `coverage.status.project.{objects,repo,refs}` with the same `target:` values + a `patch` gate at 80% for new lines + a per-component breakdown for PR comments.
- `CONTRIBUTING.md` gains a "Coverage gate" section with the threshold table, ratchet rationale for `repo`, the three-place sync requirement, and a local-run recipe.

## Closes
HeddleCo/heddle#24

## Red commit
`1c75718` — see https://github.com/HeddleCo/heddle/pull/[this-pr]/commits/1c75718. Failing tests at that SHA:

```
running 5 tests
. 1/5
tests_coverage::crate_of_extracts_top_level_crate_dir --- FAILED
.. 4/5
tests_coverage::aggregate_sums_lines_within_each_crate --- FAILED
test result: FAILED. 3 passed; 2 failed; 0 ignored; 0 measured
```

Green at `91f8374` (parser + gate logic implemented).

## Test evidence

`cargo test -p heddle-devtools` on `91f8374`:

```
running 5 tests
.....
test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

`cargo clippy -p heddle-devtools --all-targets -- -D warnings` on `91f8374`: clean.

End-to-end smoke test against the lcov.info produced by main's latest `Coverage` job (commit `3b1e8d2`) with the same `--gate` flags CI uses (`fda83eb`):

```
$ cargo run -p heddle-devtools --quiet -- audit-coverage lcov.info \
    --gate objects=80 --gate repo=78 --gate refs=80
OK   objects              lines   7127/8789   =  81.09%  (>= 80.00%)
OK   repo                 lines  11224/14269  =  78.66%  (>= 78.00%)
OK   refs                 lines   1577/1858   =  84.88%  (>= 80.00%)

audit-coverage: 3 crate(s) at or above their per-crate line-coverage threshold.
```

Failure path (same lcov, `repo=80` instead of `78`):

```
OK   objects              lines   7127/8789   =  81.09%  (>= 80.00%)
FAIL repo                 lines  11224/14269  =  78.66%  (>= 80.00%)
OK   refs                 lines   1577/1858   =  84.88%  (>= 80.00%)

audit-coverage: 1 crate(s) below threshold:
  repo: 78.66% < 80.00%
Error: audit-coverage failed
``` (exit code 1 — gate catches the regression)

## Coverage delta
| Crate | Coverage on main | Gate threshold | Goal |
|---|---|---|---|
| `objects` | 81.09% | **80%** | 80% |
| `repo` | 78.66% | **78%** (ratchet floor) | 80% |
| `refs` | 84.88% | **80%** | 80% |

`repo` sits at 78.66% on main, so the gate is wired as a ratchet floor at 78% (no-regression). Issue title's stated target is 80% — recommend follow-up issue to bring `repo` to ≥80% with new tests, then bump all three numbers in one PR (workflow + codecov.yml + CONTRIBUTING table). Surfaced before implementation in [issue comment](https://github.com/HeddleCo/heddle/issues/24#issuecomment-4464441347) — no orchestrator redirect, defaulted to ratchet approach.

## Perf delta
N/A — gate step is a single `cargo run` over the lcov.info file already produced by the previous step. Sub-second on the existing CI runners.

## Manual verification
- ✅ `cargo test -p heddle-devtools` passes at HEAD (5/5).
- ✅ `cargo clippy -p heddle-devtools --all-targets -- -D warnings` clean.
- ✅ `python3 -c "import yaml; yaml.safe_load(open(p))"` validates both `.github/workflows/rust-tests.yml` and `codecov.yml`.
- ✅ Smoke-tested gate against main's actual lcov.info (3b1e8d2): pass and fail paths both behave correctly.
- ✅ CONTRIBUTING claims grounded: `Coverage` job exists in workflow (line 71), `audit-coverage` subcommand exists (`crates/devtools/src/main.rs:15`), the new step (`Coverage gate (objects / repo / refs)`) runs before `Upload coverage to Codecov` (line 107 vs. 115).

## Blocked by → resolved
None.

## Notes for the orchestrator
- File scope as posted in [the SCOPE comment](https://github.com/HeddleCo/heddle/issues/24#issuecomment-4464441347) — no expansion.
- The `repo` 78% threshold is the only deviation from a strict reading of the issue title (\"80% for objects/repo/refs\"). The threshold is wired as a literal value in three places (workflow, codecov.yml, CONTRIBUTING table) so a follow-up PR raising it is mechanical. Recommend filing `HeddleCo/heddle#<new>`: `tests(repo): raise heddle-repo line coverage from 78.66% to 80% to clear coverage gate ratchet`.

<!-- codesmith:footer -->
---
<a href="https://app.blacksmith.sh/HeddleCo/codesmith/heddle/pr/46"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg"><img alt="View in Codesmith" src="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"></picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews
<!-- /codesmith:footer -->

### CODE REVIEW

File: .github/workflows/rust-tests.yml [hunk 1]
```diff
@@ -100,6 +100,18 @@ jobs:
...
+      # Per-crate coverage gate. Keep the --gate thresholds in sync with
+      # codecov.yml's `coverage.status.project.<crate>.target` and with
+      # the table in CONTRIBUTING.md. See heddle-devtools::run_audit_coverage
+      # for the parsing and reporting logic.
+      - name: Coverage gate (objects / repo / refs)
+        run: |
+          cargo run -p heddle-devtools --quiet -- \
+            audit-coverage lcov.info \
+              --gate objects=80 \
+              --gate repo=78 \
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply .github/workflows/rust-tests.yml]: Fixed in 95d3f45 + 0e... (typo follow-on): bumped the gate from `repo=78` to `repo=78.66` in all four sync points (rust-tests.yml, codecov.yml, CONTRIBUTING.md table + 2 code-block examples). The docstring already noted the 78.66 figure but the gate value itself hadn't been calibrated to it. Now no-regression-from-current.

File: crates/devtools/src/main.rs [hunk 2]
```diff
@@ -12,11 +12,272 @@ fn main() -> Result<()> {
...
+}
+
+/// Given an absolute or repo-relative path from an lcov `SF:` record,
+/// return the owning workspace crate name (i.e. the directory under
+/// `crates/`). Returns `None` for files outside `crates/`.
+fn crate_of(path: &str) -> Option<String> {
+    let normalized = path.replace('\\', "/");
+    let needle = "crates/";
+    let idx = normalized.find(needle)?;
+    let rest = &normalized[idx + needle.len()..];
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/devtools/src/main.rs]: Fixed in 95d3f45: switched `crate_of` from `find` to `rfind`. Now `/home/user/crates/heddle/crates/repo/src/lib.rs` correctly resolves to `repo` (the workspace member) rather than the parent-directory false match. Added a docstring note explaining the rationale.

**Comment by thethorne48** [comment]
@codex please review the updates here — addressed both findings in `95d3f45` + `5cbc071`. P2 crate_of: switched `find` → `rfind` so deeper-nested `crates/` dirs no longer false-match. P2 ratchet floor: bumped `repo` gate from 78 → 78.66 in all 4 sync points (rust-tests.yml, codecov.yml, CONTRIBUTING.md table + 2 examples + the doc-comment example I missed in the first pass). Now no-regression-from-current.

File: crates/devtools/src/main.rs [hunk 3]
```diff
@@ -12,11 +12,277 @@ fn main() -> Result<()> {
...
+/// Uses the *last* occurrence of `crates/` rather than the first, so a
+/// path like `/home/user/crates/heddle/crates/repo/src/lib.rs` resolves
+/// to `repo` (the workspace member), not `heddle` (a parent directory
+/// that happens to be named `crates`).
+fn crate_of(path: &str) -> Option<String> {
+    let normalized = path.replace('\\', "/");
+    let needle = "crates/";
+    let idx = normalized.rfind(needle)?;
+    let rest = &normalized[idx + needle.len()..];
+    let end = rest.find('/')?;
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/devtools/src/main.rs]: Fixed in <new-sha>: switched `crate_of` to path-segment-aware matching (split on `/`, find last exact `"crates"` segment with non-empty successor). My round-1 `rfind("crates/")` fix had the regression you flagged — `mycrates/foo.rs` would match inside `mycrates/`. Added regression tests for both the false-match rejection (`some_crates/`, `mycrates/`) and the deepest-exact-segment case (nested-checkout paths).

File: codecov.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        paths:
+          - crates/refs/
+
+    patch:
+      # New-or-changed lines in a PR must clear 80% — this is the
+      # forward-going floor; existing-line drift is caught by the
+      # per-crate project gates above.
+      default:
+        target: 80%
+        threshold: 1%
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply codecov.yml]: Fixed in <new-sha>: changed `threshold: 1%` → `threshold: 0%` under `patch.default`. You're right that Codecov interprets `threshold` as permitted-drop tolerance from `target`, so `80%` + `1%` was effectively a 79% gate. Added an inline comment in codecov.yml explaining the semantics so the same divergence doesn't recur.

File: crates/devtools/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/devtools/src/main.rs]: Sorry — placeholder slipped through; the SHA is **`7cf88ce`** (the previous reply had `<new-sha>` literally). Same fix described.

File: codecov.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (thethorne48) [reply codecov.yml]: Sorry — placeholder slipped through; the SHA is **`7cf88ce`** (the previous reply had `<new-sha>` literally). Same fix described.

**Comment by thethorne48** [comment]
@codex please review the updates here — addressed both round-2 findings in `7cf88ce`. P2 path-segment boundary: switched `crate_of` to split-on-`/` + last exact `crates` segment match (rejects `mycrates/`/`some_crates/` substring matches; tests added for both false-rejection and nested-checkout cases). P2 patch threshold: changed `threshold: 1%` → `0%` so the 80% target is enforced strictly.

File: crates/devtools/src/main.rs [hunk 5]
```diff
@@ -12,11 +12,302 @@ fn main() -> Result<()> {
...
+/// correctly picks the deepest workspace-member segment for nested
+/// checkouts like `/home/user/crates/heddle/crates/repo/src/lib.rs`.
+fn crate_of(path: &str) -> Option<String> {
+    let normalized = path.replace('\\', "/");
+    let segments: Vec<&str> = normalized.split('/').collect();
+    // Walk segments right-to-left; the last exact "crates" segment with
+    // a non-empty successor is the workspace-member parent.
+    for i in (0..segments.len().saturating_sub(1)).rev() {
+        if segments[i] == "crates" && !segments[i + 1].is_empty() {
+            return Some(segments[i + 1].to_string());
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/devtools/src/main.rs]: Fixed in c29359c: switched `crate_of` to workspace-member-shape matching — only `crates/<name>/<role>/...` where `<role>` is `src`/`tests`/`benches`/`examples` counts. `crates/repo/src/crates/mod.rs` now correctly returns `repo` because the inner `crates` segment lacks a `<name>/<role>/` triple after it.

Self-audit per Rule 7 explicitly: this is round 3 on the same function. Round-1 `find` → wrong-direction. Round-2 `rfind` → substring matches inside dirs. Round-3 `split + last exact crates` → inner-crates-dir matches. Round-4 (this fix) requires the role-dir suffix, which is the first variant that's stable across all the edge cases I can construct. Tests added for each previously-broken input.

**Comment by thethorne48** [comment]
@codex please review the updates here — addressed the round-3 finding in `c29359c`. Switched `crate_of` to workspace-member-shape matching: `crates/<name>/<role>/...` where `<role>` is `src`/`tests`/`benches`/`examples`. Now stable across all the edge cases (substring matches, inner-`crates`-dirs, nested checkouts, generated-file paths). Tests added for each previously-broken input. Per Rule 7, this is round 3 on the same function and the shape-match should be the stable variant.
